src/metabase/search/impl.clj

```diff
@@ -247,14 +244,263 @@
+  (let [query (:query (searchable-data-query-for-model model))]
+    (sql.helpers/where query [:= :id id])))
+
+(defn search-all-models-query [search-ctx]
+  {:select [:*]
+   :from   :search
+   :where
+   (into
+    [:or]
+    (for [model ["card"]] ; just card for now
+      (search-filters-for-model model search-ctx)))})
+
+;; -------------- Backend-specific code -----------------
+
+;; union all the queries for each model, selecting all columns from each query
+(def models ["table" "card"])
+(def search-table-name "search")
+
+;; TODO: move this to metabase.driver
+(defmulti ^:private search-type->database-type
+  "Given a search type, return the database type for that column. Applies to postgres for now"
+  {:arglists '([driver])}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
+
+;; TODO: implement for MySQL and H2
+(defmethod search-type->database-type :postgres
+  [_driver search-type]
+  (case search-type
+    :text      "TEXT"
+    :integer   "INTEGER"
+    :boolean   "BOOLEAN"
+    :timestamp "TIMESTAMP"
+    :json      "JSON"
+    :decimal   "DECIMAL"
+    :float     "DOUBLE"
+    :time      "TIME"))
+
+;; TODO: implement this for each search backend
+(defn reindex []
+  ;; 1. Create the search table with the appropriate columns
+  (let [driver  :postgres
+        columns (->> models
+                     (map #(set (:columns (searchable-data-query-for-model %))))
+                     (reduce set/union)
+                     sort)
+        columns+types (for [column columns
+                            :let [type (get (assoc search.config/all-search-columns :search_model :text)
+                                            column)]]
+                        [column (keyword (search-type->database-type driver type))])]
+    (t2/query {:drop-table [:if-exists (keyword search-table-name)]})
+    (t2/query {:create-table (keyword search-table-name)
+               :with-columns columns+types})
+  ;; 2. For each model, insert the data into the search table
+  (doseq [model models]
+    (let [{:keys [query columns]} (searchable-data-query-for-model model)
+          ->sql #(first (mdb.query/compile %))]
+      (t2/query (str (->sql {:insert-into (keyword search-table-name)
+                             :columns columns})
+                     "\n"
+                     (->sql {:select columns
+                             :from   [[query :alias]]})))))))
+
+(comment
+  (reindex)
+  ;; TODO: make this a test checking that there are no query parameters in each query
+  (assert (every? nil? (for [model ["card" "table"]]
+                         (let [{:keys [query]} (searchable-data-query-for-model model)
+                               format-sql #(sql/format % :quoted true, :dialect (sql.qp/quote-style driver))]
+                           (second (format-sql query)))))
+          )
+  )
+
+;; TODO: implement this for each search backend
+(defn index [& changes]
```

Comment:
If you split this function in the one that calls `searchable-data-query-for-instance` and then the other (the other being a multimethod) doing `insert/delete/update`, then we'll be almost there. 😁

If you change `searchable-data-query-for-instance` to `searchable-data-for-instance` - we'll be right there. 👌

But reindex has to change for that obviously... I wonder what's the difference in performance going to be 😁

Reply:
Yes to all the above. To summarize what we discussed, we can delay this change until later